proj-xcode/PowerAuth2/PowerAuthSDK.m

```diff
@@ -285,33 +285,83 @@ - (NSData*) deviceRelatedKey
-            [authentication.context invalidate];
+    //
+    // LAContext is available on this platform
+    //
+    __block NSData *key = nil;
+    __block OSStatus status;
+    BOOL executed = [PowerAuthKeychain tryLockBiometryAndExecuteBlock:^{
+        key = [_biometryOnlyKeychain dataForKey:_biometryKeyIdentifier status:&status authentication:authentication];
+    }];
+    if (key) {
+        // Key has been successfully retrieved.
+        status = errSecSuccess;
+    } else if (!executed) {
+        // Failed to acquire biometric lock. Simulate cancel in this case.
+        status = errSecUserCanceled;
+    }
+    if (status != errSecSuccess) {
+        NSError * localError;
+        PowerAuthLog(@"ERROR: Getting key for biometric authentication failed with OSStatus = %@.", @(status));
+        // The key was not fetched, try to translate OSStatus to a reasonable meaning.
+        if (status == errSecUserCanceled) {
+            // User canceled the operation.
+            localError = PA2MakeError(PowerAuthErrorCode_BiometryCancel, nil);
+        } else if (status == errSecItemNotFound) {
+            // Biometric key was not found.
+            // Note, that previously we treated this as an authentication error, but this might be
+            // an issue in application logic. For example, if app try to authenticate and immediately
+            // remove the biometry key.
+            localError = PA2MakeError(PowerAuthErrorCode_BiometryFailed, @"Biometric key not found");
+        } else if (status == errSecInvalidContext) {
+            // Invalid LAContext provided.
+            // Be aware that this code is generated in our keychain impl. Don't be confused with the naming,
+            // if LAContext is already invalidated, then general `errSecAuthFailed` is returned.
+            localError = PA2MakeError(PowerAuthErrorCode_BiometryFailed, @"Invalid LAContext");
+        } else if (status == errSecUnimplemented) {
+            // PowerAuthKeychainAuthentication was provided on platform that doesn't support it.
+            // This may happen only if tvOS application proactively create biometric key in the biometry keychain.
+            // In regular and expected setup, accessing biometry protected item on tvOS fails with errSecItemNotFound.
+            localError = PA2MakeError(PowerAuthErrorCode_BiometryFailed, @"PowerAuthKeychainAuthentication not supported");
+        } else {
+            localError = nil;
         }
-#endif
-        return key;
-    } else {
-        // Item is not present in keychain.
-        *status = errSecItemNotFound;
-        return nil;
+        // If localError variable is set, then we need to report an error.
+        if (localError) {
+            if (error) { *error = localError; }
+            return nil;
+        }
+        // No error generated, so create a fake biometry key to fail on the server.
+        key = [self generateInvalidBiometricKey];
+    } else if (error) {
```

Comment:
`if (error)` then success? Is this correct?

Reply:
Yes, that's correct. This branch happens when `status == errSecSuccess`, so I would like to set error to nil. But error is optional pointer to NSError object. It would be more readable as:
```
if (status != errSecSuccess) {
   ...
} else {
   if (error) {
       *error = nil;
   }
}

```